django/core/validators.py

```diff
@@ -86,7 +86,14 @@ class URLValidator(RegexValidator):
     hostname_re = r'[a-z' + ul + r'0-9](?:[a-z' + ul + r'0-9-]{0,61}[a-z' + ul + r'0-9])?'
     # Max length for domain name labels is 63 characters per RFC 1034 sec. 3.1
     domain_re = r'(?:\.(?!-)[a-z' + ul + r'0-9-]{1,63}(?<!-))*'
-    tld_re = r'\.(?:[a-z' + ul + r']{2,63}|xn--[a-z0-9]{1,59})\.?'
+    tld_re = (
+        '\.'                                # dot
+        '(?!-)'                             # can't start with a dash
+        '(?:[a-z-' + ul + ']{2,63}'         # domain label
```

Comment:
My point wasn't the r prefix (I just copied that from above), it was moving the dash next to the close-bracket. But now that you mentioned it -- yes, the first and last (`'\.'` and `'\.?'`) need an r prefix, because without it the strings don't have a backslash in them and these expressions will just match anything.
I think a test for this could use some invalid punctuation as the separator for the tld -- e.g. `http://unquoted~dot!`